src/model/validation/location.ts

```diff
@@ -0,0 +1,170 @@
+                value: new ProjectSource(this.sourceName, ''),
+            });
+        }
+    }
+
+    static from(location: LocationLike | undefined): MessageLocation | undefined {
+        if (!location) {
+            return undefined;
+        }
+        if (location instanceof MessageLocation) {
+            return location;
+        }
+        if (isASTNode(location)) {
+            if (!location.loc) {
+                return undefined;
+            }
+            return MessageLocation.fromGraphQLLocation(location.loc);
+        }
+        return MessageLocation.fromGraphQLLocation(location);
+    }
+
+    static fromGraphQLLocation(loc: Location) {
+        return new MessageLocation(
+            ProjectSource.fromGraphQLSource(loc.source) || loc.source.name,
+            new SourcePosition(loc.start, loc.startToken.line, loc.startToken.column),
+            new SourcePosition(
+                loc.end,
+                loc.endToken.line,
+                loc.endToken.column + loc.endToken.end - loc.endToken.start,
+            ),
+        );
+    }
+
+    get start(): SourcePosition {
+        if (this._start instanceof SourcePosition) {
+            return this._start;
+        }
+        if (!this.source) {
+            throw new Error('Can not calculate start position without reference to source');
+        }
+        const lineAndColumn = getLineAndColumnFromPosition(this._start, this.source.body);
+
+        return { offset: this._start, line: lineAndColumn.line, column: lineAndColumn.column };
+    }
+
+    get end(): SourcePosition {
+        if (this._end instanceof SourcePosition) {
+            return this._end;
+        }
+        if (!this.source) {
+            throw new Error('Can not calculate end position without reference to source');
+        }
+        const lineAndColumn = getLineAndColumnFromPosition(this._end, this.source.body);
+
+        return { offset: this._end, line: lineAndColumn.line, column: lineAndColumn.column };
+    }
+
+    toString() {
+        return `${this.sourceName}:${this.start.line}:${this.start.column}`;
+    }
+}
+
+function isASTNode(obj: LocationLike): obj is ASTNode {
+    return 'kind' in obj;
+}
+
+export function locationWithinStringArgument(
+    node: LocationLike,
+    offset: number,
+    length: number,
+): MessageLocation | undefined {
+    if (isASTNode(node) && node.kind === Kind.STRING && node.loc) {
+        const loc = node.loc;
+        // add 1 because of "
+        return new MessageLocation(
+            ProjectSource.fromGraphQLSource(loc.source) || loc.source.name,
+            new SourcePosition(
+                loc.start + 1 + offset,
+                loc.startToken.line,
+                loc.startToken.column + 1 + offset,
+            ),
+            new SourcePosition(
+                loc.start + 1 + offset + length,
+                loc.endToken.line,
+                loc.startToken.column + 1 + offset + length,
+            ),
```

Comment:
Should the end location not be corrected by -1 since there is also an `"` at the end of the argument?

I probably do not understand this 100% and this is correct since this code was already present before.

Reply:
the end location is still based on `loc.start`, not on `loc.end`, so the trailing `"` does not matter.

(It's still not 100% correct because it does not account for escaped character sequence, but it's good enough for our use cases)